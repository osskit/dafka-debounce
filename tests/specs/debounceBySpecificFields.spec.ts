import delay from 'delay';

import type {Orchestrator} from '../testcontainers/orchestrator.js';
import {start} from '../testcontainers/orchestrator.js';
import {produce} from '../services/produce.js';
import {getOffset} from '../services/getOffset.js';
import {consume} from '../services/consume.js';

const groupId = 'test11';
const sourceTopic = 'test-input';
const targetTopic = 'test-output';

describe('tests', () => {
    let orchestrator: Orchestrator;

    beforeEach(
        async () => {
            orchestrator = await start(
                {
                    KAFKA_BROKER: 'kafka:9092',
                    MONITORING_SERVER_PORT: '3000',
                    GROUP_ID: groupId,
                    SOURCE_TOPIC: sourceTopic,
                    TARGET_TOPIC: targetTopic,
                    WINDOW_DURATION: '5',
                    OUTPUT_PARTITIONING_KEY: 'SPECIFIC_FIELDS',
                    OUTPUT_PARTITIONING_KEY_FIELDS: '/data/type',
                },
                [sourceTopic, targetTopic]
            );
        },
        5 * 60 * 1000
    );

    afterEach(async () => {
        if (!orchestrator) {
            return;
        }
        await orchestrator.stop();
    });

    it('should debounce by specific fields', async () => {
        await produce(orchestrator, {topic: sourceTopic, messages: [{value: JSON.stringify({data: {type: 'bar'}}), key: '1'}]});
        await produce(orchestrator, {topic: sourceTopic, messages: [{value: JSON.stringify({data: {type: 'bar'}}), key: '2'}]});

        await delay(5000);

        await expect(consume(orchestrator.kafkaClient, targetTopic)).resolves.toMatchSnapshot();

        await delay(5000);

        await expect(getOffset(orchestrator.kafkaClient, groupId, sourceTopic)).resolves.toBe(2);
    });
});
