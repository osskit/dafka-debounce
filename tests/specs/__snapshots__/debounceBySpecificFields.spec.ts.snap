// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests should debounce by specific fields 1`] = `
{
  "headers": {},
  "value": {
    "data": {
      "type": "bar",
    },
  },
}
`;
